MATH-1648: Derivative computations for BicubicInterpolator

amoscatelli opened the pull request.

aherbert commented on commons-math-legacy/src/test/java/org/apache/commons/math4/legacy/analysis/interpolation/BicubicInterpolatingFunctionTest.java:
```diff
@@ -304,7 +304,7 @@ public void testSplinePartialDerivatives() {
         }
 
-        final BicubicFunction f = new BicubicFunction(coeff, true);
+        final BicubicFunction f = new BicubicFunction(coeff, 1, 1, true);
```
Have you tested this when `xR` and `yR` are not 1? I would suggest simply repeating the test with `xR` and `yR` as  0.5 and 2 then testing the derivative is correctly scaled.

aherbert replied:
That uses an interval of 1 / 20 for both axes. Perhaps update to have different step sizes for each axis. This will ensure xR and yR are not mixed up.

amoscatelli replied:
Done ! I updated that test to have different xval and yval array, and it's still working.

aherbert commented on commons-math-legacy/src/test/java/org/apache/commons/math4/legacy/analysis/interpolation/BicubicInterpolatingFunctionTest.java:
```diff
@@ -286,6 +287,231 @@ public double value(double x, double y) {
     }
+    
+    /**
+     * Test for partial derivatives of {@link BicubicSplineFunction}.
```
`BicubicSplineFunction` does not exist; it should be `BicubicFunction`

amoscatelli replied:
Done!

aherbert commented on commons-math-legacy/src/test/java/org/apache/commons/math4/legacy/analysis/interpolation/BicubicInterpolatingFunctionTest.java:
```diff
@@ -286,6 +287,231 @@ public double value(double x, double y) {
+
+    /**
+     * Test that the partial derivatives computed from a
+     * {@link BicubicSplineInterpolatingFunction} match the input data.
```
`BicubicSplineInterpolatingFunction` does not exist

amoscatelli replied:
Done!

aherbert commented on commons-math-legacy/src/main/java/org/apache/commons/math4/legacy/analysis/interpolation/BicubicInterpolator.java:
```diff
@@ -41,6 +41,30 @@
+
+    /**
+     * Default constructor.
+     * The argument {@link #BicubicSplineInterpolator(boolean) initializeDerivatives}
```
`BicubicSplineInterpolator` does not exist

amoscatelli replied:
Done!

aherbert commented on commons-math-legacy/src/main/java/org/apache/commons/math4/legacy/analysis/interpolation/BicubicInterpolator.java:
```diff
@@ -41,6 +41,30 @@
+     *
+     * @param initializeDerivatives Whether to initialize the internal data
+     * needed for calling any of the methods that compute the partial derivatives
+     * of the {@link BicubicSplineInterpolatingFunction function} returned from
```
`BicubicSplineInterpolatingFunction` does not exist

amoscatelli replied:
Done!

aherbert commented on commons-math-legacy/src/main/java/org/apache/commons/math4/legacy/analysis/interpolation/BicubicInterpolatingFunction.java:
```diff
@@ -181,6 +226,75 @@ public boolean isValidPoint(double x, double y) {
+    /**
+     * @return the first partial derivative respect to x.
+     * @throws NullPointerException if the internal data were not initialized
+     * (cf. {@link #BicubicSplineInterpolatingFunction(double[],double[],double[][],
```
`BicubicSplineInterpolatingFunction` does not exist. Please update all 5 references in the javadoc of this class to `BicubicInterpolatingFunction`

amoscatelli replied:
Done!

aherbert reviewed (changes requested):
Thanks for the addition.

There are some javadoc errors due to referencing the old class names with `Spline`.

I do not understand why there was a switch to the matrix multiplication. The original `computeSplineCoefficients` would be much more efficient. IIUC it should be a simple row vs column multiplication.

amoscatelli commented on commons-math-legacy/src/main/java/org/apache/commons/math4/legacy/analysis/interpolation/BicubicInterpolatingFunction.java:
```diff
@@ -317,49 +325,6 @@ private int searchIndex(double c, double[] val) {
-
-        for (int i = 0; i < NUM_COEFF; i++) {
-            double result = 0;
-            final double[] row = AINV[i];
```
I thought using the Matrix functionality would have made the code much clearer than reinventing the wheel.
If you want I can rollback it to the previous method (that was indeed multiplying the matrix with the column) but I would expect the matrix packages to be efficient ...

aherbert replied:
If you find the actual method in `org.apache.commons.math4.legacy.linear.Array2DRowRealMatrix.multiply` then you will see that it is not efficient at all. In my experience using the OO matrix classes is 5-10x slower than a manually written implementation that avoids defensive copying of all intermediate arrays.

Please revert the change as it is not required (functionality is the same).

amoscatelli replied:
I feel kinda sad.
Anyway I reverted as you suggested !

amoscatelli commented on commons-math-legacy/src/main/java/org/apache/commons/math4/legacy/analysis/interpolation/BicubicInterpolatingFunction.java:
```diff
@@ -322,4 +485,40 @@ private double apply(double[] pX, double[] pY, double[][] coeff) {
+    /**
+     * @return the partial derivative wrt {@code x}.
+     */
+    public BivariateFunction partialDerivativeX() {
```
Done!

amoscatelli commented on commons-math-legacy/src/main/java/org/apache/commons/math4/legacy/analysis/interpolation/BicubicInterpolator.java:
```diff
@@ -41,6 +41,30 @@
 public class BicubicInterpolator
     implements BivariateGridInterpolator {
+    /** Whether to initialize internal data used to compute the analytical
+    derivatives of the splines. */
```
Done!

amoscatelli commented:
@aherbert 
I reverted the matrix computation and fixed the test to have different xval and yval arrays.

This should be good !
Please let me know if there is something more I can do

aherbert commented:
OK. I have authorized the CI build. Hopefully it is all clean.

amoscatelli commented:
@aherbert 
I see errors ... but they seem kinda unrelated to this PR ...

aherbert commented:
The simplex optimizer tests are not very robust. I will try restarting to see if they complete on a second attempt.

amoscatelli commented:
Better, but that's still the simplex optimizer

amoscatelli commented:
@aherbert what needs to be done to make it progress ?

amoscatelli commented:
@aherbert strange, it says there is still 1 review / change pending ... but every one has been resolved ...

aherbert commented:
Some of the unit tests use random seeds. As such any objects used by these tests may have coverage variations due to the code execution paths differing between runs. What is important here is the diff coverage is 100%, i.e. what you added is all covered by tests.

The changes look good. Please see the Jira ticket for a suggestion to improve efficiency of the computation.

aherbert commented:
Closed by #8e6dee595832951e4d74e795a76e5aff236d4181
